roachtest: sqlsmith/setup=empty/setting=no-mutations failed

[(roachtest).sqlsmith/setup=empty/setting=no-mutations failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2062841&tab=buildLog) on [master@456a07cfc1e53b87abc7709052e54efb1450e758](https://github.com/cockroachdb/cockroach/commits/456a07cfc1e53b87abc7709052e54efb1450e758):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/sqlsmith/setup=empty/setting=no-mutations/run_1
	sqlsmith.go:169,sqlsmith.go:199,test_runner.go:757: error: pq: internal error: inconsistent Case return types oid unknown
		stmt:
		SELECT
			tab_4599.col_7948 AS col_7949, tab_4598.col_7946 AS col_7950, 3072052637:::OID AS col_7951
		FROM
			(VALUES ('X':::STRING), (COALESCE(e'\'':::STRING, NULL)), (NULL), (NULL)) AS tab_4598 (col_7946)
			JOIN (
					VALUES
						(CASE WHEN true THEN 2505127111:::OID ELSE COALESCE(NULL, NULL) END, '':::STRING),
						(0:::OID, e'z\x16)4Z%':::STRING),
						(1328346255:::OID, '':::STRING)
				)
					AS tab_4599 (col_7947, col_7948) ON (tab_4598.col_7946) = (tab_4599.col_7948)
		ORDER BY
			tab_4598.col_7946 ASC, tab_4599.col_7947;

```

<details><summary>More</summary><p>


Artifacts: [/sqlsmith/setup=empty/setting=no-mutations](https://teamcity.cockroachdb.com/viewLog.html?buildId=2062841&tab=artifacts#/sqlsmith/setup=empty/setting=no-mutations)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Asqlsmith%2Fsetup%3Dempty%2Fsetting%3Dno-mutations.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/sql/logictest/testdata/logic_test/typing
+++ b/pkg/sql/logictest/testdata/logic_test/typing
@@ -237,3 +237,9 @@ SELECT 1::pg_catalog.special_int
 # don't have types.
 query error pq: type "crdb_internal.mytype" does not exist
 SELECT 1::crdb_internal.mytype
+
+# Regression test for #50978.
+query T
+SELECT CASE WHEN true THEN 1234:::OID ELSE COALESCE(NULL, NULL) END
+----
+1234

--- a/pkg/sql/opt/exec/execbuilder/scalar.go
+++ b/pkg/sql/opt/exec/execbuilder/scalar.go
@@ -101,6 +101,10 @@ func (b *Builder) buildTypedExpr(
 }
 
 func (b *Builder) buildNull(ctx *buildScalarCtx, scalar opt.ScalarExpr) (tree.TypedExpr, error) {
+	if scalar.DataType().Family() == types.TupleFamily {
+		// See comment in buildCast.
+		return tree.DNull, nil
+	}
 	return tree.ReType(tree.DNull, scalar.DataType()), nil
 }
 
@@ -321,6 +325,13 @@ func (b *Builder) buildCast(ctx *buildScalarCtx, scalar opt.ScalarExpr) (tree.Ty
 	input, err := b.buildScalar(ctx, cast.Input)
 	if err != nil {
 		return nil, err
+	}
+	if cast.Typ.Family() == types.TupleFamily {
+		// TODO(radu): casts to Tuple are not supported (they can't be serialized
+		// for distsql). This should only happen when the input is always NULL so
+		// the expression should still be valid without the cast (though there could
+		// be cornercases where the type does matter).
+		return input, nil
 	}
 	return tree.NewTypedCastExpr(input, cast.Typ), nil
 }

--- a/pkg/sql/opt/norm/testdata/rules/scalar
+++ b/pkg/sql/opt/norm/testdata/rules/scalar
@@ -136,7 +136,7 @@ project
  ├── scan a
  │    └── columns: i:2
  └── projections
-      └── COALESCE(i:2, NULL, NULL) [as=coalesce:7, outer=(2)]
+      └── COALESCE(i:2, CAST(NULL AS INT8), CAST(NULL AS INT8)) [as=coalesce:7, outer=(2)]
 
 norm expect=SimplifyCoalesce
 SELECT COALESCE((1, 2, 3), (2, 3, 4)) FROM a

--- a/pkg/sql/opt/optbuilder/scalar.go
+++ b/pkg/sql/opt/optbuilder/scalar.go
@@ -253,8 +253,13 @@ func (b *Builder) buildScalar(
 
 	case *tree.CoalesceExpr:
 		args := make(memo.ScalarListExpr, len(t.Exprs))
+		typ := t.ResolvedType()
 		for i := range args {
-			args[i] = b.buildScalar(t.TypedExprAt(i), inScope, nil, nil, colRefs)
+			// The type of the CoalesceExpr might be different than the inputs (e.g.
+			// when they are NULL). Force all inputs to be the same type, so that we
+			// build coalesce operator with the correct type.
+			expr := tree.ReType(t.TypedExprAt(i), typ)
+			args[i] = b.buildScalar(expr, inScope, nil, nil, colRefs)
 		}
 		out = b.factory.ConstructCoalesce(args)
 

--- a/pkg/sql/opt/optbuilder/testdata/scalar
+++ b/pkg/sql/opt/optbuilder/testdata/scalar
@@ -1243,3 +1243,19 @@ if-err [type=decimal]
  │    └── variable: "@1":1 [type=decimal]
  └── err-code
       └── const: '10000' [type=string]
+
+# Verify that we build a Coalesce of the correct type when the inputs
+# are NULL (#50978).
+build-scalar
+CASE WHEN true THEN 1234:::OID ELSE COALESCE(NULL, NULL) END
+----
+case [type=oid]
+ ├── true [type=bool]
+ ├── when [type=oid]
+ │    ├── true [type=bool]
+ │    └── const: 1234 [type=oid]
+ └── coalesce [type=oid]
+      ├── cast: OID [type=oid]
+      │    └── null [type=unknown]
+      └── cast: OID [type=oid]
+           └── null [type=unknown]

--- a/pkg/sql/physicalplan/expression.go
+++ b/pkg/sql/physicalplan/expression.go
@@ -120,9 +120,10 @@ func (e *evalAndReplaceSubqueryVisitor) VisitPre(expr tree.Expr) (bool, tree.Exp
 			e.err = err
 			return false, expr
 		}
-		var newExpr tree.Expr = val
-		if _, isTuple := val.(*tree.DTuple); !isTuple && expr.ResolvedType().Family() != types.UnknownFamily {
-			newExpr = tree.NewTypedCastExpr(val, expr.ResolvedType())
+		newExpr := tree.Expr(val)
+		typ := expr.ResolvedType()
+		if _, isTuple := val.(*tree.DTuple); !isTuple && typ.Family() != types.UnknownFamily && typ.Family() != types.TupleFamily {
+			newExpr = tree.NewTypedCastExpr(val, typ)
 		}
 		return false, newExpr
 	default:

